From ffd3f410ef4c6a3e68853471c874775922f48beb Mon Sep 17 00:00:00 2001
From: Hanjun Guo <guohanjun@huawei.com>
Date: Wed, 29 Sep 2021 16:02:07 +0800
Subject: [PATCH] usb: xhci: Add workaround for phytium
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: ffd3f410ef4c6a3e68853471c874775922f48beb
Modified-by-SEL: Yes, refreshed due to context changes


hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I4CGER
CVE: NA

---------------------------

Add workaround for phytium as the firmware didn't report the
DMA size info.

Signed-off-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Reviewed-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/usb/host/xhci-pci.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

--- a/drivers/usb/host/xhci-pci.c
+++ b/drivers/usb/host/xhci-pci.c
@@ -437,6 +437,18 @@ static int xhci_pci_update_hub_device(st
 	return xhci_update_hub_device(hcd, hdev, tt, mem_flags);
 }
 
+#ifdef CONFIG_ARM64
+#include <asm/cputype.h>
+static void phytium_xhci_pci_workaround(struct pci_dev *dev)
+{
+	/* Firmware bug, DMA mask is not reported by the firmware */
+	if (read_cpuid_implementor() == ARM_CPU_IMP_PHYTIUM)
+		dma_set_mask(&dev->dev, DMA_BIT_MASK(64));
+}
+#else
+static inline void phytium_xhci_pci_workaround(struct pci_dev *dev) { }
+#endif
+
 /*
  * We need to register our own PCI probe function (instead of the USB core's
  * function) in order to create a second roothub under xHCI.
@@ -461,6 +473,8 @@ static int xhci_pci_probe(struct pci_dev
 		return PTR_ERR(reset);
 	reset_control_reset(reset);
 
+	phytium_xhci_pci_workaround(dev);
+
 	/* Prevent runtime suspending between USB-2 and USB-3 initialization */
 	pm_runtime_get_noresume(&dev->dev);
 
